src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -0,0 +1,347 @@
+      reorganiseOrEnqueue(block, bestBlock, weight)
+    }
+    importResult.foreach(measureBlockMetrics)
+    importResult
+  }
+
+  private def doBlockPreValidation(block: Block)(implicit
+      blockchainConfig: BlockchainConfig
+  ): Task[Either[ValidationBeforeExecError, BlockExecutionSuccess]] =
+    Task
+      .evalOnce {
+        val validationResult = blockValidation.validateBlockBeforeExecution(block)
+        validationResult.left.foreach { error =>
+          log.error("Error while validating block with hash {} before execution: {}", block.hash, error.reason)
+        }
+        validationResult
+      }
+      .executeOn(validationScheduler)
+
+  private def returnNoTotalDifficulty(bestBlock: Block): Task[BlockImportFailed] = {
+    log.error(
+      "Getting total difficulty for current best block with hash: {} failed",
+      bestBlock.header.hashAsHexString
+    )
+    Task.now(
+      BlockImportFailed(
+        s"Couldn't get total difficulty for current best block with hash: ${bestBlock.header.hashAsHexString}"
+      )
+    )
+  }
+
+  private def returnNoBestBlock(): Task[BlockImportFailed] = {
+    log.error("Getting current best block failed")
+    Task.now(BlockImportFailed("Couldn't find the current best block"))
+  }
+
+  private def isBlockADuplicate(block: BlockHeader, currentBestBlockNumber: BigInt): Boolean = {
+    val hash = block.hash
+    blockchainReader.getBlockByHash(hash).isDefined && block.number <= currentBestBlockNumber || blockQueue.isQueued(
+      hash
+    )
+  }
+
+  private def isPossibleNewBestBlock(newBlock: BlockHeader, currentBestBlock: BlockHeader): Boolean =
+    newBlock.number == currentBestBlock.number + 1 && newBlock.parentHash == currentBestBlock.hash
+
+  private def measureBlockMetrics(importResult: BlockImportResult): Unit =
+    importResult match {
+      case BlockImportedToTop(blockImportData) =>
+        blockImportData.foreach(blockData => BlockMetrics.measure(blockData.block, blockchainReader.getBlockByHash))
+      case ChainReorganised(_, newBranch, _) =>
+        newBranch.foreach(block => BlockMetrics.measure(block, blockchainReader.getBlockByHash))
+      case _ => ()
+    }
+
+  private def importToTop(
+      block: Block,
+      currentBestBlock: Block,
+      currentWeight: ChainWeight
+  )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[BlockImportResult] =
+    Task
+      .evalOnce(importBlockToTop(block, currentBestBlock.header.number, currentWeight))
+      .executeOn(blockExecutionScheduler)
+
+  /** *
+    * Open for discussion: this is code that was in BlockImport. Even thought is tested (before) that only one block
+    * is being added to the blockchain, this code assumes that the import may be of more than one block (a branch)
+    * and because it is assuming that it may be dealing with a branch the code to handle errors assumes several scenarios.
+    * Is there is reason for this over-complication?
+    */
```

Comment:
I added this comment because I think this deserves a discussion! What are your thoughts?

Reply:
I don't know the history to be able to know why it was like that. But in our case we will probably want to import several blocks in one run, so I think we might run into a similar scenario where only a part of the branch executes. 
But I agree that in that case it does not really make sense as we check that the block is the direct child of the current best block 🤔.
I think we can keep it as it is right now, as we know we will rewrite this part to fit our design anyway.

Reply:
```scala
    val executionResult = for {
      topBlock <- blockQueue.enqueueBlock(block, bestBlockNumber)
      topBlocks = blockQueue.getBranch(topBlock.hash, dequeue = true)
      (executed, errors) = blockExecution.executeAndValidateBlocks(topBlocks, currentWeight)
    } yield (executed, errors, topBlocks)
```

pretty sure the assumption is that this line:
`blockQueue.getBranch(topBlock.hash, dequeue = true)`
can return a branch that has not been stored or executed.

this is the internal caching that BlockImport is doing.